cmd/sfe/main.go

```diff
@@ -0,0 +1,226 @@
+package notmain
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/config"
+	"github.com/letsencrypt/boulder/features"
+	bgrpc "github.com/letsencrypt/boulder/grpc"
+	blog "github.com/letsencrypt/boulder/log"
+	rapb "github.com/letsencrypt/boulder/ra/proto"
+	sapb "github.com/letsencrypt/boulder/sa/proto"
+	"github.com/letsencrypt/boulder/sfe"
+)
+
+type Config struct {
+	SFE struct {
+		DebugAddr string `validate:"omitempty,hostname_port"`
+
+		// ListenAddress is the address:port on which to listen for incoming
+		// HTTP requests. Defaults to ":80".
+		ListenAddress string `validate:"omitempty,hostname_port"`
+
+		// TLSListenAddress is the address:port on which to listen for incoming
+		// HTTPS requests. If none is provided the SFE will not listen for HTTPS
+		// requests.
+		TLSListenAddress string `validate:"omitempty,hostname_port"`
+
+		// Timeout is the per-request overall timeout. This should be slightly
+		// lower than the upstream's timeout when making request to the SFE.
+		Timeout config.Duration `validate:"-"`
+
+		// ShutdownStopTimeout is the duration that the SFE will wait before
+		// shutting down any listening servers.
+		ShutdownStopTimeout config.Duration
+
+		// AllowOrigins is for setting CORS on OPTIONS requests
+		AllowOrigins []string
+
+		ServerCertificatePath string `validate:"required_with=TLSListenAddress"`
+		ServerKeyPath         string `validate:"required_with=TLSListenAddress"`
+
+		TLS cmd.TLSConfig
+
+		RAService *cmd.GRPCClientConfig
+		SAService *cmd.GRPCClientConfig
+
+		Unpause struct {
+			// Seed is a secret that should contain 256 bits (32 bytes) of
+			// random data used to derive an x/crypto/ed25519 keypair (e.g. the
+			// output of `openssl rand -hex 16`). In a multi-DC deployment this
+			// value should be the same across all boulder-wfe and sfe
+			// instances.
+			Seed cmd.PasswordConfig `validate:"-"`
+		}
```

Comment:
This should move to `cmd/config.go` so we can use it in the WFE as well. See my comment in `sfe/sfe.go`.